crates/wassette/src/component_storage.rs

```diff
@@ -0,0 +1,248 @@
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                tracing::debug!(component_id = %component_id, path = %path.display(), "{} already absent", description);
+            }
+            Err(e) => {
+                return Err(anyhow!(
+                    "Failed to remove {} at {}: {}",
+                    description,
+                    path.display(),
+                    e
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    /// Create a validation stamp for a component artifact to track stale data on disk.
+    ///
+    /// When `include_hash` is `true` the SHA-256 hash of the file is
+    /// recorded in addition to size and modification time so changes can be
+    /// detected even when timestamps are unreliable.
+    pub async fn create_validation_stamp(
+        path: &Path,
+        include_hash: bool,
+    ) -> Result<ValidationStamp> {
+        let metadata = tokio::fs::metadata(path)
+            .await
+            .with_context(|| format!("Failed to read metadata for {}", path.display()))?;
+
+        let file_size = metadata.len();
+        let mtime = metadata
+            .modified()
+            .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            .and_then(|t| {
+                t.duration_since(std::time::UNIX_EPOCH)
+                    .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            })?
+            .as_secs();
+
+        let content_hash = if include_hash {
+            let bytes = tokio::fs::read(path)
+                .await
+                .with_context(|| format!("Failed to read {} for hashing", path.display()))?;
+            let mut hasher = sha2::Sha256::new();
+            hasher.update(&bytes);
+            Some(format!("{:x}", hasher.finalize()))
+        } else {
+            None
+        };
+
+        Ok(ValidationStamp {
+            file_size,
+            mtime,
+            content_hash,
+        })
+    }
+
+    /// Check if the validation stamp matches the current file on disk.
+    pub async fn validate_stamp(path: &Path, stamp: &ValidationStamp) -> bool {
+        let Ok(metadata) = tokio::fs::metadata(path).await else {
+            return false;
+        };
+
+        if metadata.len() != stamp.file_size {
+            return false;
+        }
+
+        let Ok(mtime) = metadata
+            .modified()
+            .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            .and_then(|t| {
+                t.duration_since(std::time::UNIX_EPOCH)
+                    .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            })
+            .map(|d| d.as_secs())
+        else {
+            return false;
+        };
+
+        if mtime != stamp.mtime {
+            return false;
+        }
+
+        if let Some(expected_hash) = &stamp.content_hash {
```

Comment:
Because we are always checking the hash if it is set, we should probably just shortcut by doing this first. If the hash hasn't changed, then we can be pretty certain we're ok and we don't event need to do the time check